src/backend/api/meals.js

```diff
@@ -3,13 +3,149 @@ const router = express.Router();
 const knex = require("../database");
 
 router.get("/", async (request, response) => {
-  try {
-    // knex syntax for selecting things. Look up the documentation for knex for further info
-    const titles = await knex("meals").select("title");
-    response.json(titles);
-  } catch (error) {
-    throw error;
+    //maxPrice
+    if("maxPrice" in request.query){
+      const maxPrice = Number(request.query.maxPrice);
+      if(isNaN(maxPrice)){
+        response.send('maxPrice should be a number!')
+      }else{
+         meals = knex("meal").where("price", "<", maxPrice);
+      }
+    }
+     //availableReservations
+    if("availableReservations" in request.query){
+      const reqAvailReservation = request.query.availableReservations;
+      if(reqAvailReservation != 'true'){
+        response.send('enter true value of availableReservations!')
+      }else {
+        meals = knex.raw(`select meal.id , meal.title, (meal.max_reservations)-(reservation.number_of_guests) AS AvailableReservation from meal inner join reservation on meal.id = reservation.meal_id where ((max_reservations)-(number_of_guests)) > 0`)
+      }
+    }
+     //title
+    if("title" in request.query) {
+      const reqTitle = request.query.title;
+      if(!isNaN(reqTitle)){
+        response.send("Title should be a string! ")
+      }else{
+        meals = knex('meal').where("title", "like", `%${reqTitle}%`)
+      }
+    }
+    //dateAfter
+    if("dateAfter" in request.query){
+      const reqDate = new Date(request.query.dateAfter)
+      meals =  knex('meal').where("when_date", ">", reqDate )
+    }
+    //dateBefore
+    if("dateBefore" in request.query){
+      const dateReq = new Date(request.query.dateBefore)
+      meals =  knex('meal').where("when_date", "<", dateReq  )
+    }
+    //limit
+    if("limit" in request.query){
+      const limit = Number(request.query.limit)
+      if(isNaN(request.query.limit)){
+        response.send('enter a number!')
+      }else{
+        meals =  knex('meal').limit(limit)
+      }
+    }
+    //api/meals?sort_key=price&sort_dir=desc together
+    if("sort_key" in request.query){
+      const sortKey = request.query.sort_key;
+      const sortDir = request.query.sort_dir;
+      const array = ["price", "when_date", "max_reservations"]
+      array.forEach( arr =>{
+        if (sortKey == arr && sortDir == "desc"){
+          meals = knex('meal').orderBy(arr, "desc")
+        }else
+         if(sortKey == arr){
+          meals = knex('meal').orderBy(arr)
+        }
+      })
+    }
+    //api/meals?sort_key=price&sort_dir=desc
+  try{
+      const allMeals = await meals;
+      if(allMeals.length === 0){
+        response.send([]);
+      }else{
+        response.send(allMeals);
+      }
```

Comment:
Why do you need to deal with `allMeals` being empty? Couldn't you just do `response.send(allMeals)`?